### PR TITLE
Ghost and Race demo time extraction fix

### PR DESCRIPTION
**Pull request by Redix**
On systems that use a language with comma as decimal-point character, the decimal places of the ghosts and race demos are dropped. The sscanf call stops when it reaches the dot used by the servers.

The time parsing functions are probably a bit more complicated then needed, but they are able to parse the chat messages of any ddnet, ddrace and race version I encountered.
Seconds are parsed up to the third decimal place.

### CODE REVIEW

File: src/game/client/components/race.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	// these functions return the time in milliseconds, time 0 is invalid
+
+	static int TimeFromSecondsStr(const char *pStr) // x.xxx
+	{
+		int Time = str_toint(pStr) * 1000;
+		while(*pStr >= '0' && *pStr <= '9') pStr++;
+		if(*pStr == '.' || *pStr == ',')
+		{
+			pStr++;
+			const int Mult[3] = { 100, 10, 1 };
```
Comment (heinrich5991) [inline src/game/client/components/race.h]: `static`? `MULT` (or `s_aMult`).

File: src/game/client/components/race.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+			for(int i = 0; pStr[i] >= '0' && pStr[i] <= '9' && i < 3; i++)
+				Time += (pStr[i] - '0') * Mult[i];
+		}
+		return Time;
+	}
+
+	static int TimeFromStr(const char *pStr) // x minute(s) x.xx second(s)
+	{
+		static const char *pMinutesStr = " minute(s) ";
+		static const char *pSecondsStr = " second(s)";
```
Comment (heinrich5991) [inline src/game/client/components/race.h]: `static const char *const SECONDS_STR` (or `s_pSecondsStr`).

File: src/game/client/components/race.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+class CRaceHelper
+{
+public:
+	// these functions return the time in milliseconds, time 0 is invalid
+
+	static int TimeFromSecondsStr(const char *pStr) // x.xxx
+	{
+		int Time = str_toint(pStr) * 1000;
+		while(*pStr >= '0' && *pStr <= '9') pStr++;
```
Comment (heinrich5991) [inline src/game/client/components/race.h]: newline

File: src/game/client/components/race.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+			return str_toint(pStr) * 60 * 1000 + TimeFromSecondsStr(pMinutes + str_length(pMinutesStr));
+		else
+			return TimeFromSecondsStr(pStr);
+	}
+
+	static int TimeFromFinishMessage(const char *pStr, char *pNameBuf, int NameBufSize) // xxx finished in: x minute(s) x.xx second(s)
+	{
+		static const char *pFinishedStr = " finished in: ";
+		const char *pFinished = str_find(pStr, pFinishedStr);
+		int FinishedPos = pFinished - pStr;
```
Comment (heinrich5991) [inline src/game/client/components/race.h]: This is undefined behavior if `pFinished` is a null pointer.

**Comment by Redix** [comment]
I use 0 as invalid because str_toint also returns 0 if it cannot parse the string. When using -1, I would need to check if str_toint returned 0 because the string was really a representation of 0 or invalid.

I would prefer to use integers for all times but even in the ghost files times are stored as floats (one reason why race mod uses a new format). So this would mean changing quite a lot of code.

**Comment by Redix** [comment]
Invalid is -1 now and the code only uses integers for time handling (except for the time value inside the ghost files).

**Comment by heinrich5991** [comment]
Awesome!

Thanks a lot for your pull request. :) CI failure on Mac is caused by something else.

**Pull request merged**